src/CloudNative.CloudEvents.NewtonsoftJson/JsonEventFormatter.cs

```diff
@@ -136,6 +136,44 @@ public override CloudEvent DecodeStructuredModeMessage(Stream data, ContentType
         public override CloudEvent DecodeStructuredModeMessage(byte[] data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes) =>
             DecodeStructuredModeMessage(new MemoryStream(data), contentType, extensionAttributes);
 
+        public override async Task<IReadOnlyList<CloudEvent>> DecodeBatchModeMessageAsync(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes)
+        {
+            data = data ?? throw new ArgumentNullException(nameof(data));
+
+            var jsonReader = CreateJsonReader(data, contentType.GetEncoding());
+            var array = await JArray.LoadAsync(jsonReader).ConfigureAwait(false);
+            return DecodeJArray(array, extensionAttributes);
+        }
+
+        public override IReadOnlyList<CloudEvent> DecodeBatchModeMessage(Stream data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes)
+        {
+            data = data ?? throw new ArgumentNullException(nameof(data));
+
+            var jsonReader = CreateJsonReader(data, contentType.GetEncoding());
+            var array = JArray.Load(jsonReader);
+            return DecodeJArray(array, extensionAttributes);
+        }
+
+        public override IReadOnlyList<CloudEvent> DecodeBatchModeMessage(byte[] data, ContentType contentType, IEnumerable<CloudEventAttribute> extensionAttributes) =>
+            DecodeBatchModeMessage(new MemoryStream(data), contentType, extensionAttributes);
+
+        private IReadOnlyList<CloudEvent> DecodeJArray(JArray jArray, IEnumerable<CloudEventAttribute> extensionAttributes = null)
+        {
+            List<CloudEvent> events = new List<CloudEvent>();
+            foreach (var token in jArray)
+            {
+                if (token is JObject obj)
+                {
+                    events.Add(DecodeJObject(obj, extensionAttributes));
+                }
+                else
+                {
+                    throw new ArgumentException("Invalid array element within batch", "data");
```

Comment:
consider passing in paramName

Reply:
Should we include the token in the exception?

Reply:
Happy to pass in paramName to avoid the hard-coding, yes - thanks.
I don't want to include the actual token, as generally I don't like exceptions containing potentially arbitrary data (both in terms of size and PII)... but we could include the type.

Reply:
Right.. I forgot about that regarding including arbitrary data. We can also include the index of the failing element.

Reply:
Done.